packages/segmented-control/src/SegmentedControl.tsx

```diff
@@ -0,0 +1,314 @@
+  border-style: solid;
+  background-color: var(--indicator-background-color);
+  border-color: var(--indicator-border-color);
+`;
+
+const interactionRingStyle = css`
+  z-index: 3;
+`;
+
+/**
+ * Types
+ */
+export interface SegmentedControlProps {
+  children: React.ReactNode;
+  name?: string;
+  size?: Size;
+  darkMode?: boolean;
+  defaultValue?: string;
+  value?: string;
+  onChange?: React.ChangeEventHandler<HTMLInputElement>;
+  className?: string;
+  label?: string;
+}
+
+/**
+ * Component
+ */
+const SegmentedControl = React.forwardRef<
+  HTMLDivElement,
+  SegmentedControlProps
+>(function SegmentedControl(
+  {
+    children,
+    name: nameProp,
+    size = 'default',
+    darkMode = false,
+    defaultValue = '',
+    value: controlledValue,
+    onChange,
+    className,
+    label,
+    ...rest
+  }: SegmentedControlProps,
+  forwardedRef,
+) {
+  // TODO log warning if defaultValue is set but does not match any child value
+
+  const [getRef, setRef] = useDynamicRefs<HTMLInputElement>();
+
+  const mode = darkMode ? 'dark' : 'light';
+
+  // If a value is given, then it's controlled
+  let isControlled = controlledValue != null ? true : false;
+
+  React.Children.forEach(children, (child, i) => {
+    if (isComponentType(child, 'SegmentedControlOption')) {
+      // If one of the options has been programatically checked, then it's controlled
+      if (child.props.checked != null) {
+        isControlled = true;
+      }
+
+      // If there's no default value given, then we fall back to the first option
+      if (!defaultValue && i == 0) {
+        defaultValue = child.props.value;
+      }
+    }
+  });
+
+  // Keep track of the uncontrolled value
+  const [uncontrolledValue, setUncontrolledValue] = useState<string>(
+    defaultValue,
+  );
+
+  const name = useIdAllocator({ prefix: 'radio-group', id: nameProp });
+
+  // Change handler
+  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (onChange) {
+      onChange(e);
+    }
+
+    if (!isControlled) {
+      setUncontrolledValue(e.target.value);
+    }
+  };
+
+  // Add internal props to children passed in
+  const renderedChildren = React.Children.map(children, (child, index) => {
+    if (!isComponentType(child, 'SegmentedControlOption')) {
+      console.warn(`${child} is not a SegmentedControlOption`);
+      return child;
+    }
+
+    const _checked: boolean = isControlled
+      ? child.props.value === controlledValue || !!child.props.checked
+      : child.props.value === uncontrolledValue;
+
+    return React.cloneElement(child, {
+      id: child.props.id || `${name}-${index}`,
+      darkMode,
+      size,
+      _checked,
+      _name: name,
+      _onChange: handleChange,
+      ref: setRef(`${name}-${index}`),
+    });
+  });
+
+  const selectedIndex = React.Children.toArray(
+    renderedChildren,
+  ).findIndex(child =>
+    isControlled
+      ? (child as React.ReactElement).props.value === controlledValue
```

Comment:
I _think_ we can type child without typecasting here by typing the output array from `React.Children.toArray`.